torchao/prototype/autoround/multi_tensor.py

```diff
@@ -0,0 +1,172 @@
+        shape = kwargs.pop("shape", input.shape)
+        return torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)
+
+    def __init__(self, input, **kwargs):
+        self.values = []
+        self.count = 0
+        self.add_tensors(input)
+        self.debug = True
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}(data={self.values})"
+
+    def add_tensors(self, input):
+        if isinstance(input, (tuple, list)):
+            for inp in input:
+                self.add_tensors(inp)
+        else:
+            assert isinstance(
+                input, torch.Tensor
+            ), f"MultiTensor can only use add_tensors for Tensors or lists of tensors but got {type(input)}"
+            self.count += 1
+            self.values.append(input)
+        return self
+
+    def pad_to_length(self, length):
+        if self.count > length:
+            return self
+        self.add_tensors([self.values[-1]] * (length - self.count))
+        return self
+
+    @classmethod
+    def flat_to_grouped(cls, flat):
+        # size of biggest MultiTensor
+        multi_tensor_size = max(
+            [x.count if isinstance(x, MultiTensor) else 1 for x in flat]
+        )
+        # convert [A, MultiTensor(b1,b2,b3), MultiTensor(c1,c2,c3)] => [[A,b1,c1], [A,b2,c2] [A,b3,c3]]
+        grouped = list(
+            zip(
+                *[
+                    (
+                        x.pad_to_length(multi_tensor_size).values
+                        if isinstance(x, MultiTensor)
+                        else [x] * multi_tensor_size
+                    )
+                    for x in flat
+                ]
+            )
+        )
+        return grouped
+
+    @classmethod
+    def grouped_to_flat(cls, grouped):
+        # convert [[A,b1,c1], [A,b2,c2] [A,b3,c3]] => [A, MultiTensor(b1,b2,b3), MultiTensor(c1,c2,c3)]
+        # where A is nontensor, b's,c's are tensors
+        # convert [[A,b1,c1], [A,b2,c2] [A,b3,c3]] => [(A,A,A), (b1,b2,b3), (c1,c2,c3)]
+        flat_tups = list(zip(*grouped))
+        # convert [(A,A,A), (b1,b2,b3), (c1,c2,c3)] => [A, MultiTensor(b1,b2,b3), MultiTensor(c1,c2,c3)]
+        flattened = [
+            cls(tup) if isinstance(tup[0], torch.Tensor) else tup[0]
+            for tup in flat_tups
+        ]
+        # need to check that getting rid of all but one from each nonTensor tuple is OK
+        non_tensors_equal = min(
+            [True]
+            + [
+                min(
+                    [True]
+                    + [  # handle situation where tuples have size 0
+                        tup[0] == x for x in tup  # check all elements match
+                    ]
+                )
+                for tup in flat_tups
+                if not isinstance(tup[0], torch.Tensor)  # look at tuples of nonTensors
+            ]
+        )
+        return flattened, non_tensors_equal
+
+    @classmethod
+    def revert_to_tensor_pairs(cls, args, kwargs=None):
```

Comment:
Hi @jerryzh168, I added this class method to revert the `args` and `kwargs` to their original native tensor pairs. These native tensor pairs are used by `func` at `__torch_function__` and `auto-round` side.

Reply:
will it be easier if you implement `__iter__` for multi tensor: https://github.com/pytorch/ao/blob/31ff4d9649fe35f175552ce428d1fe7ff212fc19/tutorials/calibration_flow/gptq_like.py#L72